utils/src/test/java/software/amazon/awssdk/utils/cache/NonBlockingTest.java

```diff
@@ -0,0 +1,82 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.cache;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import org.junit.jupiter.api.MethodOrderer;
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestMethodOrder;
+import org.mockito.Mockito;
+import org.mockito.stubbing.Answer;
+
+// These tests assert on the largest pool size, so the test order matters (smaller expected sizes before larger expected sizes)
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
+public class NonBlockingTest {
+    @Test
+    @Order(1)
+    public void threadsAreSharedBetweenNonBlockingInstances() throws InterruptedException {
+        List<NonBlocking> nbs = new ArrayList<>();
+        try {
+            // Create 99 concurrent non-blocking instances
+            for (int i = 0; i < 99; i++) {
+                NonBlocking nb = new NonBlocking("test", Duration.ofMillis(100));
+                nb.initializeCachedSupplier(Mockito.mock(CachedSupplier.class));
+                nbs.add(nb);
+                Thread.sleep(10);
+            }
+
+            Thread.sleep(1_000);
+
+            // Make sure we used less-than 99 to do the refreshes.
+            assertThat(NonBlocking.executor().getLargestPoolSize()).isLessThan(99);
+        } finally {
+            nbs.forEach(NonBlocking::close);
+        }
+    }
+
+    @Test
+    @Order(2)
+    public void refreshesAreMaxed() throws InterruptedException {
+        CachedSupplier<?> slowSupplier = Mockito.mock(CachedSupplier.class);
+        Mockito.when(slowSupplier.get()).thenAnswer((Answer<Void>) invocation -> {
+            Thread.sleep(1_000);
+            return null;
+        });
+
+        List<NonBlocking> nbs = new ArrayList<>();
+        try {
+            for (int i = 0; i < 1_000; i++) {
+                NonBlocking nb = new NonBlocking("test", Duration.ofMillis(0));
+                nb.initializeCachedSupplier(slowSupplier);
+                nbs.add(nb);
+            }
+
+            Thread.sleep(1_000);
+
+            // In a perfect world this would be capped to 100, but the mechanism we use to limit concurrent refreshes usually
+            // means more than 100 get created. 150 should be a reasonable limit to check for, because without the limiter it
+            // would be ~1000.
+            assertThat(NonBlocking.executor().getLargestPoolSize()).isLessThanOrEqualTo(150);
```

Comment:
is 150 the highest possible value  ? or can we have 151 because current mechanisms to limit concurrent refreshes

Reply:
90% of my experiments were < 100, and 100% were < 105. 

The reality is that it could be as many as the number of non-blocking instances we have (in this situation, 1000), but it's exceedingly unlikely under a fair scheduler. It would require EVERY thread to release the rate limiting lock, and then get parked while the task scheduler woke up (after ~60 seconds), acquired the rate limiting lock that got released, and then create a new thread to execute the task.

In reality, the scheduler would never park a thread for 60 seconds. The only reason we see it happening occasionally here is because we decreased the time between tasks to 0ms to be able to test more quickly.